File: fli/search/_wire.py
r"""Parsing helpers for Google Flights' FlightsFrontendService wire format.

The Service returns JSONP-flavoured responses of the form::

    )]}'\n\n
    <chunk1_byte_len>\n
    [["wrb.fr", null, "<inner JSON string>"]]
    <chunk2_byte_len>\n
    [["wrb.fr", null, "<inner JSON string>"]]
    ...

`GetShoppingResults` and `GetCalendarGraph` happen to emit a single chunk so
the legacy parsers in this package could get away with `lstrip(")]}'")`.
`GetBookingResults` emits two chunks, so we need a proper multi-chunk reader.

Important quirk: the length headers count UTF-8 **bytes**, not Python string
characters. When the response contains any non-ASCII characters (which it
sometimes does — airport names, airline names) the offsets diverge, so the
reader must operate over the byte representation of the body.

This module centralises that reader and exposes :func:`iter_wrb_chunks` which
yields the decoded inner JSON of each ``wrb.fr`` chunk.
"""

from __future__ import annotations

import json
import logging
from collections.abc import Iterator
from typing import Any

logger = logging.getLogger(__name__)

_PREFIX = b")]}'"


def iter_wrb_chunks(body: str | bytes) -> Iterator[Any]:
    """Yield the inner JSON object of every ``wrb.fr`` chunk in ``body``.

    Robust to single-chunk responses with no length headers (the older
    ``GetShoppingResults`` / ``GetCalendarGraph`` shape) — those are parsed
    by falling back to a single JSON load over the trimmed body.
    """
    if isinstance(body, str):
        raw = body.encode("utf-8")
    else:
        raw = body

    raw = raw.lstrip()
    if raw.startswith(_PREFIX):
        raw = raw[len(_PREFIX) :]
    raw = raw.lstrip()

    if not raw:
        return

    # Fast path: no length headers (legacy single-chunk responses).
    if not (b"0" <= raw[:1] <= b"9"):
        try:
            outer = json.loads(raw.decode("utf-8"))
        except (ValueError, json.JSONDecodeError, UnicodeDecodeError):
            logger.warning("Failed to decode single-chunk wrb.fr body as JSON", exc_info=True)
            return
        yield from _chunks_from_outer(outer)
        return

    cursor = 0
    while cursor < len(raw):
        # Read the decimal length prefix terminated by \n.
        end = raw.find(b"\n", cursor)
        if end == -1:
            break
        try:
            length = int(raw[cursor:end])
        except ValueError:
            logger.warning(
                "Malformed length header at offset %d; truncating chunk stream",
                cursor,
            )
            break
        # Google's length header counts the leading newline after the header
        # AND the trailing newline that separates this chunk from the next.
        # We've already consumed the leading newline (it terminated the header),
        # so we read `length - 1` bytes which gives JSON + trailing \n.
        cursor = end + 1
        chunk_bytes = max(length - 1, 0)
        payload = raw[cursor : cursor + chunk_bytes]
        cursor += chunk_bytes
        try:
            outer = json.loads(payload.strip().decode("utf-8"))
        except (ValueError, json.JSONDecodeError, UnicodeDecodeError):
            logger.warning("Discarding malformed wrb.fr chunk", exc_info=True)
            continue
        yield from _chunks_from_outer(outer)


def _chunks_from_outer(outer: Any) -> Iterator[Any]:
    """Walk a top-level chunk list and yield decoded inner-JSON payloads."""
    if not isinstance(outer, list):
        return
    for row in outer:
        if not isinstance(row, list) or len(row) < 3:
            continue
        if row[0] != "wrb.fr":
            continue
        inner = row[2]
        if not isinstance(inner, str) or not inner:
            continue
        try:
            yield json.loads(inner)
        except (ValueError, json.JSONDecodeError):
            logger.warning("Failed to decode wrb.fr inner JSON payload", exc_info=True)
            continue


def parse_first_wrb_payload(body: str | bytes) -> Any:
    """Return the inner JSON of the first ``wrb.fr`` chunk, or None."""
    for chunk in iter_wrb_chunks(body):
        return chunk
    return None

File: fli-js/src/search/wire.ts
/**
 * Parser for Google Flights' FlightsFrontendService wire format.
 *
 * 1:1 port of fli/search/_wire.py — handles both the legacy single-chunk
 * JSONP shape and the multi-chunk format used by GetBookingResults.
 *
 * Important quirk: the length headers count UTF-8 BYTES, not characters,
 * so the parser operates on `Uint8Array` rather than the JS string.
 */

const PREFIX = ")]}'";
const PREFIX_BYTES = new TextEncoder().encode(PREFIX);

const decoder = new TextDecoder("utf-8", { fatal: false });

function toBytes(body: string | Uint8Array): Uint8Array {
  if (body instanceof Uint8Array) return body;
  return new TextEncoder().encode(body);
}

function lstrip(buf: Uint8Array): Uint8Array {
  let i = 0;
  while (i < buf.length) {
    const b = buf[i];
    if (b !== 0x20 && b !== 0x09 && b !== 0x0a && b !== 0x0d) break;
    i++;
  }
  return buf.subarray(i);
}

function startsWith(buf: Uint8Array, prefix: Uint8Array): boolean {
  if (buf.length < prefix.length) return false;
  for (let i = 0; i < prefix.length; i++) {
    if (buf[i] !== prefix[i]) return false;
  }
  return true;
}

function indexOfByte(buf: Uint8Array, byte: number, from = 0): number {
  for (let i = from; i < buf.length; i++) {
    if (buf[i] === byte) return i;
  }
  return -1;
}

function* chunksFromOuter(outer: unknown): Generator<unknown> {
  if (!Array.isArray(outer)) return;
  for (const row of outer) {
    if (!Array.isArray(row) || row.length < 3) continue;
    if (row[0] !== "wrb.fr") continue;
    const inner = row[2];
    if (typeof inner !== "string" || inner.length === 0) continue;
    try {
      yield JSON.parse(inner);
    } catch {
      // Skip malformed inner JSON.
    }
  }
}

/** Yield the inner JSON of every `wrb.fr` chunk in `body`. */
export function* iterWrbChunks(body: string | Uint8Array): Generator<unknown> {
  let raw = toBytes(body);
  raw = lstrip(raw);
  if (startsWith(raw, PREFIX_BYTES)) {
    raw = raw.subarray(PREFIX_BYTES.length);
  }
  raw = lstrip(raw);
  if (raw.length === 0) return;

  // Fast path: legacy single-chunk responses with no length headers.
  const first = raw[0];
  if (first === undefined || first < 0x30 || first > 0x39) {
    try {
      const outer = JSON.parse(decoder.decode(raw));
      yield* chunksFromOuter(outer);
    } catch {
      // Discard malformed body.
    }
    return;
  }

  let cursor = 0;
  while (cursor < raw.length) {
    const end = indexOfByte(raw, 0x0a, cursor);
    if (end === -1) break;
    const headerText = decoder.decode(raw.subarray(cursor, end));
    // Python's `int(...)` raises on `"12abc"`; `Number.parseInt` returns 12,
    // which would slip the parser into garbage chunk offsets. Require a pure
    // decimal header to keep wire-format strictness in sync.
    if (!/^[0-9]+$/.test(headerText)) break;
    const length = Number.parseInt(headerText, 10);
    if (!Number.isFinite(length)) break;
    cursor = end + 1;
    const chunkBytes = Math.max(length - 1, 0);
    const payload = raw.subarray(cursor, cursor + chunkBytes);
    cursor += chunkBytes;
    try {
      const trimmed = decoder.decode(payload).trim();
      const outer = JSON.parse(trimmed);
      yield* chunksFromOuter(outer);
    } catch {
      // Discard malformed chunks.
    }
  }
}

/** Return the inner JSON of the first `wrb.fr` chunk, or `null`. */
export function parseFirstWrbPayload(body: string | Uint8Array): unknown {
  for (const chunk of iterWrbChunks(body)) {
    return chunk;
  }
  return null;
}
